crates/ruff_linter/src/rules/refurb/rules/math_constant.rs

```diff
@@ -0,0 +1,88 @@
+use crate::checkers::ast::Checker;
+use crate::importer::ImportRequest;
+use anyhow::Result;
+use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, Number};
+use ruff_text_size::Ranged;
+
+/// ## What it does
+/// Checks for literals that are similar to constants in `math` module.
+///
+/// ## Why is this bad?
+/// Hard coding mathematical constants like π increases code duplication, reduces readability and
+/// may lack precision.
+///
+/// ## Example
+/// ```python
+/// A = 3.141592 * r**2
+/// ```
+///
+/// Use instead:
+/// ```python
+/// A = math.pi * r**2
+/// ```
+///
+/// ## References
+/// - [Python documentation: `math` constants](https://docs.python.org/3/library/math.html#constants)
+#[violation]
+pub struct MathConstant {
+    literal: String,
+    constant: &'static str,
+}
+
+impl Violation for MathConstant {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let MathConstant { literal, constant } = self;
+        format!("Replace `{literal}` with `math.{constant}`")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let MathConstant { constant, .. } = self;
+        Some(format!("Use `math.{constant}`"))
+    }
+}
+
+fn convert_to_constant(
+    literal: &ast::ExprNumberLiteral,
+    constant: &'static str,
+    checker: &Checker,
+) -> Result<Fix> {
+    let (edit, binding) = checker.importer().get_or_import_symbol(
+        &ImportRequest::import("math", constant),
+        literal.start(),
+        checker.semantic(),
+    )?;
+    Ok(Fix::unsafe_edits(
+        Edit::range_replacement(binding, literal.range()),
+        [edit],
+    ))
```

Comment:
Why is this an unsafe fix?

Reply:
Most likely it changes result of calculations. For example, `math.pi` is more precise than `3.14`.

Reply:
I'm unsure if that means we need to mark it as unsafe, I could go either way though. @charliermarsh?

If we do leave it as unsafe, we should document why in the rule docs.

Reply:
Hah, this is an interesting case. I guess I'd lean towards unsafe since it _could_ break tests, etc., if you had hard-coded expectations that differed slightly?

Reply:
I guess but tests written to check floating point math already should include tolerances. Idk I'm struggling to think of a case where this would change runtime behavior in a meaningful way.

Reply:
Since it's in preview, it may be fine to just use safe and see if we get feedback.